Changing agent service installation mechanism on Linux OS (leaving forever, forever!)

NimrodGeva opened the pull request:
### Purpose

Unfortunately for forever users (e.g us), upstart is no longer supported as a service startup mechanism on newer linux distributions.
*Roughly, it goes as follows:
1. System V:
   - Debian 6 and earlier
   - Ubuntu 9.04 and earlier
   - CentOS 5 and earlier
2. Upstart:
   - Ubuntu 9.10 to Ubuntu 14.10, including Ubuntu 14.04
   - CentOS 6
3. systemd:
   - Debian 7 and Debian 8
   - Ubuntu 15.04 and newer
   - CentOS 7

The [forever-service](https://github.com/foreverjs/forever) node module had an abstraction of handling service set up, but only handled System V and Upstart. In order to not over-complicate different mechanisms for different distributions, we left forever (forever) and now use a module called [node-linux](https://github.com/coreybutler/node-linux), which only supports Upstart and systemd.

A large part of this pull request is a shift of the service installation from shell scripts to node.js.

*This is not always clear-cut, because of some intermingling (for instance, both Upstart and systemd should be on an Ubuntu 15), and the fact that the existence or lack thereof of the initialisation mechanism is not guaranteed in any of the above cases.
### Checklist
- Code:
  - [x] User Experience: **Should be unaffected**
  - [x] Failure handling and Comments on non trivial sections: 
  - [x] Cross Platform: **Supporting systems using upstart and systemd**
  - [x] Code Review: @nimrod-becker 
  - [ ] Code Review: @tamireran 
- Testing:
  - [x] Unit/System Tests: **N/A**
  - [x] Tested with: `npm test`
  - Manual testing: Checked installation, upgrade and reboot of agents on the following:
  - [ ] Ubuntu 8
  - [ ] Ubuntu 10
  - [x] Ubuntu 12
  - [x] Ubuntu 14
  - [x] Ubuntu 16
  - [ ] CentOs 5
  - [x] CentOs 6
  - [x] CentOs 7
  - [x] RHEL 7
- Supportability:
  - [x] Diagnostics info, Phone Home info, ActivityLog events, External Syslog: **N/A**
- Upgrade:
  - [x] Mongo Schema Upgrade: **N/A**
  - [x] Agents changes, Server Platform changes and New packages added to package.json: **Agent upgrades checked manually, new package added to package.json**
### Fixed Issues References
- Fixes #1392

NimrodGeva committed:
Changing linux agent install method

Changing service install and keep alive mechanisms for Ubuntu 16 and
CentOs 7 support.

NimrodGeva committed:
Merge branch 'master' into nimrodg_bug_#1392_branch

NimrodGeva committed:
Merge branch 'master' into nimrodg_bug_#1392_branch

NimrodGeva committed:
Merge branch 'master' into nimrodg_bug_#1392_branch

NimrodGeva committed:
Merge branch 'master' into nimrodg_bug_#1392_branch

NimrodGeva committed:
Merge branch 'master' into nimrodg_bug_#1392_branch

NimrodGeva committed:
install node-linux

NimrodGeva committed:
linux agent install fixes

Split uninstall and install scripts.
Now enabling systemd (for boot service start)

NimrodGeva committed:
place holder for else

NimrodGeva committed:
Fixes to agent installer

NimrodGeva committed:
Cleanups in wrapper

NimrodGeva committed:
fixes in agent wrap

NimrodGeva committed:
fixes to agent install scripts

NimrodGeva committed:
Attempted fix in service startup

NimrodGeva committed:
Tests with agent installation

NimrodGeva committed:
Initial service start workaround

NimrodGeva committed:
Attempted fix

NimrodGeva committed:
more tests

NimrodGeva committed:
losing patience :D

NimrodGeva committed:
commenting out dbg

NimrodGeva committed:
REVERT THIS LATER

NimrodGeva committed:
Added working dir for the service

NimrodGeva committed:
Added path to uninstall script too

NimrodGeva committed:
rly?

NimrodGeva committed:
fix attempt

NimrodGeva committed:
Attempt to move service start back to install

NimrodGeva committed:
Removing workarounds from agent install

NimrodGeva committed:
removed debug comment out

NimrodGeva committed:
No implicit reinstall

This change makes it so that to reinstall one must call the script with
—repair, otherwise it will not do anything if already installed.

NimrodGeva committed:
Merge branch 'master' into nimrodg_bug_#1392_branch

NimrodGeva committed:
Added upstart support

Added upstart initialisation script. Will be copied to /init whenever
on non-systemd os. A bit naive but it’s a pretty solid heuristic.

nimrod-becker reviewed (changes requested):
See my comments below

NimrodGeva committed:
A few fixes for upstart

NimrodGeva committed:
Removed node-linux from package

oops.

NimrodGeva committed:
differentiating system v

NimrodGeva committed:
Fixes to upstart script

NimrodGeva committed:
bugfix in install scripts

NimrodGeva committed:
starting service in upstart after setup

NimrodGeva committed:
bugfix

NimrodGeva committed:
bug fig

oops

NimrodGeva committed:
not attempting service install on upstart

NimrodGeva committed:
fixes for certain environments

NimrodGeva committed:
Merge branch 'master' into nimrodg_bug_#1392_branch

NimrodGeva committed:
changing init mechanism detection heuristics

also adding a default case that attempts to do everything (though this
should never happen)

NimrodGeva committed:
Code review remarks

NimrodGeva committed:
Code tidying

Finished code review fixes.
Prevented a potential bug in reinstallation of service.
Made internal agent work with the new init mechanism

NimrodGeva committed:
bugfix

oops

NimrodGeva committed:
Merge branch 'master' into nimrodg_bug_#1392_branch

NimrodGeva committed:
bugfix

this is getting embarrassing.

NimrodGeva committed:
Changed system d detection heuristic

There are many heuristics for this. One could potentially even combine
them.
This is simply an alignment with node-linux’s heuristic, to prevent
differing diagnostics.

nimrod-becker reviewed (commented):
Please add comments at the head of the files regarding what is their purpose.

See my other comments inline

NimrodGeva committed:
changed log path

NimrodGeva committed:
debugging

need to revert this later

NimrodGeva committed:
Revert "debugging"

This reverts commit cb691e6d48e46026a2f660a5d99403d3ab5db061.

NimrodGeva committed:
debugging

revert

NimrodGeva committed:
bugfix

On some systems, node-linux’s service install emitted a ‘doesnotexist’
event if it was trying to install and nothing was installed before.
If the said system happened to be using system V, we ended up
installing two services.

NimrodGeva committed:
Revert "debugging"

This reverts commit 697a80c72b2c6bb4b849cece48dd45b153650e22.

NimrodGeva committed:
Simplification of repair flow

Fix for a bug on certain situations that caused 2 agents to run.

NimrodGeva committed:
initialising on old centos machines

NimrodGeva committed:
heuristic change for upstart detection

NimrodGeva committed:
fixes to install script

Changing method for activating upstart services for centos 6.
removing forever service if it exists.

NimrodGeva committed:
change to upstart script

for some reason, events were not triggered as expected on centos 6

NimrodGeva committed:
Added a few comments

This is mostly to force jenkins to rebuild and test upgrade of agents.

NimrodGeva committed:
debugging

NimrodGeva committed:
added sleep before server re up after upgrade

NimrodGeva committed:
debugging

NimrodGeva committed:
debugging

NimrodGeva committed:
missing semicolon

NimrodGeva committed:
debugging

NimrodGeva committed:
going back to the repair flow

NimrodGeva committed:
debugging

NimrodGeva committed:
debugging

NimrodGeva committed:
debugging

NimrodGeva committed:
debugging

NimrodGeva committed:
debugging

NimrodGeva committed:
Running upgrade detached

NimrodGeva committed:
trying to prevent wrapper from killing setup

NimrodGeva committed:
another attempt

NimrodGeva committed:
debugging

NimrodGeva committed:
debugging

NimrodGeva committed:
debugging

NimrodGeva committed:
debugging

NimrodGeva committed:
debugging

NimrodGeva committed:
took ownership of system d script

NimrodGeva committed:
debugging

NimrodGeva committed:
debugging

NimrodGeva committed:
debugging

NimrodGeva committed:
debugging

NimrodGeva committed:
debugging

NimrodGeva committed:
debugging

NimrodGeva committed:
debugging

NimrodGeva committed:
debugging

NimrodGeva committed:
Merge branch 'master' into nimrodg_bug_#1392_branch

NimrodGeva committed:
reverting debug changes

building a package to test changes

NimrodGeva committed:
separated upgrade script

also some debug print cleanups.

NimrodGeva committed:
Merge branch 'master' into nimrodg_bug_#1392_branch

NimrodGeva committed:
debugging

NimrodGeva committed:
Applied changes to upstart and systemv

NimrodGeva committed:
Refactored uninstallation out of installation

NimrodGeva committed:
debugging

NimrodGeva committed:
Refactored out the removal of services

NimrodGeva committed:
Fixes to installation scripts

NimrodGeva committed:
making sure remove_service is executable

NimrodGeva committed:
Merge branch 'master' into nimrodg_bug_#1392_branch

NimrodGeva committed:
fixing lint error

NimrodGeva committed:
bugfix

The supervisor should be running the agent wrap directly instead of
letting the service mechanism take responsibility.

NimrodGeva committed:
Fixed location of upgrade script

NimrodGeva committed:
building upgrade package for test

NimrodGeva committed:
building upgrade package

NimrodGeva committed:
Merge branch 'master' into nimrodg_bug_#1392_branch
